File: rust/sdk/src/builders/document.rs
use crate::DocumentUpdate;
use core::convert::From;
use m10_protos::prost::Message;
use m10_protos::Pack;
use m10_protos::{sdk, Collection};

#[derive(Default)]
pub struct DocumentBuilder {
    docs: Vec<sdk::Operation>,
}

impl DocumentBuilder {
    pub fn insert<P: Pack>(mut self, doc: P) -> Self {
        self.docs.push(sdk::Operation::insert(doc));
        self
    }

    pub fn insert_custom(mut self, collection: Collection, doc: impl Message) -> Self {
        self.docs.push(sdk::Operation {
            operation: Some(sdk::operation::Operation::InsertDocument(
                sdk::operation::InsertDocument {
                    collection: collection.to_string(),
                    document: doc.encode_to_vec(),
                },
            )),
        });
        self
    }

    pub fn delete<P: Pack>(mut self, doc: P) -> Self {
        self.docs
            .push(sdk::Operation::delete::<P>(doc.id().to_vec()));
        self
    }

    pub fn delete_custom(mut self, collection: Collection, id: Vec<u8>) -> Self {
        self.docs.push(sdk::Operation {
            operation: Some(sdk::operation::Operation::DeleteDocument(
                sdk::operation::DeleteDocument {
                    collection: collection.to_string(),
                    primary_key: Some(bytes::Bytes::from(id).into()),
                },
            )),
        });
        self
    }

    pub fn update<P: Pack>(mut self, update: DocumentUpdate<P>) -> Self {
        self.docs.push(update.operation());
        self
    }
}

impl From<DocumentBuilder> for sdk::DocumentOperations {
    fn from(doc: DocumentBuilder) -> Self {
        Self {
            operations: doc.docs,
        }
    }
}

File: node/m10-sdk/src/builders/document.ts
import { m10 } from "../../protobufs";
import type { Collection, CollectionDocument, DocumentUpdate } from "../collections";
import { getUint8ArrayFromDocument } from "../collections";


export class DocumentBuilder {
    private _docs: m10.sdk.Operation[];


    public constructor() {
        this._docs = [];
    }

    public insert<D extends CollectionDocument>(collection: Collection, document: D): DocumentBuilder {
        this._docs.push(
            new m10.sdk.Operation({
                insertDocument: new m10.sdk.Operation.InsertDocument({
                    collection: collection,
                    document: getUint8ArrayFromDocument(collection, document),
                }),
            }),
        );
        return this;
    }

    public insertCustom(collection: Collection, document: Uint8Array): DocumentBuilder {
        this._docs.push(
            new m10.sdk.Operation({
                insertDocument: new m10.sdk.Operation.InsertDocument({
                    collection: collection,
                    document: document,
                }),
            }),
        );
        return this;
    }

    public delete<D extends CollectionDocument>(collection: Collection, document: D): DocumentBuilder {
        this._docs.push(
            new m10.sdk.Operation({
                deleteDocument: new m10.sdk.Operation.DeleteDocument({
                    collection: collection,
                    primaryKey: new m10.sdk.Value({ bytesValue: document.id }),
                }),
            }),
        );
        return this;
    }

    public deleteCustom(collection: Collection, id: Uint8Array): DocumentBuilder {
        this._docs.push(
            new m10.sdk.Operation({
                deleteDocument: new m10.sdk.Operation.DeleteDocument({
                    collection: collection,
                    primaryKey: new m10.sdk.Value({ bytesValue: id }),
                }),
            }),
        );
        return this;
    }

    public update<D extends CollectionDocument>(update: DocumentUpdate<D>): DocumentBuilder {
        this._docs.push(update.toOperation());
        return this;
    }


    public toDocumentOperations(): m10.sdk.DocumentOperations {
        return new m10.sdk.DocumentOperations({ operations: this._docs });
    }
}
